server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,351 @@
+        stats.numCollisions.inc();
+        CounterMetric numCollisions = collidedIntCounters.get(transformedValue);
+        if (numCollisions == null) { // First time the transformedValue has had a collision
+            numCollisions = new CounterMetric();
+            numCollisions.inc(2);
+            collidedIntCounters.put(transformedValue, numCollisions); // Initialize the number of colliding keys to 2
+        } else {
+            numCollisions.inc();
+        }
+    }
+
+    @Override
+    public boolean add(Integer value) {
+        if (value == null) {
+            return false;
+        }
+        stats.numAddAttempts.inc();
+
+        if (getSize() % REFRESH_SIZE_EST_INTERVAL == 0) {
+            mostRecentByteEstimate = getMemorySizeInBytes();
+        }
+        if (getMemorySizeCapInBytes() > 0 && mostRecentByteEstimate > getMemorySizeCapInBytes()) {
+            stats.atCapacity.set(true);
+            return false;
+        }
+        int transformedValue = transform(value);
+
+        writeLock.lock();
+        try {
+            boolean alreadyContained;
+            // saves calling transform() an additional time
+            alreadyContained = rbm.contains(transformedValue);
+            if (!alreadyContained) {
+                rbm.add(transformedValue);
+                stats.size.inc();
+                return true;
+            }
+            // If the value is already pending removal, take it out of the removalList
+            HashSet<Integer> removalSet = removalSets.get(transformedValue);
+            if (removalSet != null) {
+                removalSet.remove(value);
+                // Don't increment the counter - this is handled by handleCollisions() later
+                if (removalSet.isEmpty()) {
+                    removalSets.remove(transformedValue);
+                }
+            }
+
+            handleCollisions(transformedValue);
+            return false;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public boolean contains(Integer value) {
+        if (value == null) {
+            return false;
+        }
+        int transformedValue = transform(value);
+        readLock.lock();
+        try {
+            return rbm.contains(transformedValue);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Integer getInternalRepresentation(Integer value) {
+        if (value == null) {
+            return 0;
+        }
+        return Integer.valueOf(transform(value));
+    }
+
+    /**
+     * Attempts to remove a value from the keystore. WARNING: Removing keys which have not been added to the keystore
+     * may cause undefined behavior, including future false negatives!!
+     * @param value The value to attempt to remove.
+     * @return true if the value was removed, false otherwise
+     */
+    @Override
+    public boolean remove(Integer value) {
```

Comment:
This method is understandably complex, but we can make this a bit more readable ?

Reply:
I have refactored this to below code, see if u can do better than me. 
```java
    @Override
    public boolean remove(Integer value) {
        if (value == null) {
            return false;
        }
        int transformedValue = transform(value);
        if (!containsInRBM(transformedValue))
            return false;
        // move below into write lock
        stats.numRemovalAttempts.inc();
        writeLock.lock();
        try {
            CounterMetric numCollisions = collidedIntCounters.get(transformedValue);
            if (numCollisions == null) {
                // There's not been a collision for this transformedValue, so we can safely remove
                removeFromRBM(transformedValue);
                return true;
            }
            // This transformed value has had a collision before
            HashSet<Integer> removalSet = removalSets.get(transformedValue);
            if (removalSet == null) {
                // First time a removal has been attempted for this transformed value
                // Add the key value, not the transformed value, to the list of attempted removals for this transformedValue
                removalSets.put(transformedValue, new HashSet<>(value));
                numCollisions.dec();
            } else {
                if (removalSet.contains(value)) {
                    return false; // We have already attempted to remove this value. Do nothing
                }
                removalSet.add(value);
                numCollisions.dec();
                // If numCollisions has reached zero, we can safely remove all values in removalList
                if (numCollisions.count() == 0) {
                    clearValueFromAllConfigs(transformedValue);
                    return true;
                }
            }
            return false;
        } finally {
            writeLock.unlock();
        }
    }

    private void clearValueFromAllConfigs(int transformedValue) {
        if (!lock.isWriteLockedByCurrentThread()) {
            throw new IllegalStateException("Lock must be held when calling this method");
        }
        removeFromRBM(transformedValue);
        collidedIntCounters.remove(transformedValue);
        removalSets.remove(transformedValue);
    }

    private boolean containsInRBM(int transformedValue) {
        readLock.lock();
        try {
            return rbm.contains(transformedValue);
        } finally {
            readLock.unlock();
        }
    }
```